app/models/user_bookmark_list.rb

```diff
@@ -21,10 +21,42 @@ def initialize(user:, guardian:, params:)
 
   def load
     @bookmarks = BookmarkQuery.new(user: @user, guardian: @guardian, params: @params).list_all
+    if SiteSetting.use_polymorphic_bookmarks
+      preload_polymorphic_associations
+    end
     @bookmarks
   end
 
   def per_page
     @per_page ||= @params[:per_page]
   end
+
+  private
+
+  # We have already confirmed the user has access to these records at
+  # this point in BookmarkQuery, so it is safe to load them directly
+  # without any further security checks.
+  #
+  # These polymorphic associations are loaded to make the UserBookmarkListSerializer's
+  # life easier, which conditionally chooses the bookmark serializer to use based
+  # on the type, and we want the associations all loaded ahead of time to make
+  # sure we are not doing N1s.
+  def preload_polymorphic_associations
+    @topics = Topic.includes(:topic_users).where(
+      id: Bookmark.select_type(@bookmarks, "Topic").map(&:bookmarkable_id)
+    ).where(topic_users: { user_id: @user.id })
+
+    @posts = Post.includes(topic: :topic_users).where(
+      id: Bookmark.select_type(@bookmarks, "Post").map(&:bookmarkable_id)
+    ).where(topic_users: { user_id: @user.id })
+
+    Bookmark.registered_bookmarkables.each do |registered_bookmarkable|
+      bookmarkable_ids = Bookmark.select_type(@bookmarks, registered_bookmarkable.model.name).map(&:bookmarkable_id)
+      self.instance_variable_set(
+        :"@#{registered_bookmarkable.table_name}",
+        registered_bookmarkable.preload_associations(bookmarkable_ids)
+      )
+      self.class.send(:attr_reader, registered_bookmarkable.table_name)
```

Comment:
We can leverage on `ActiveRecord::Associations::Preloader` here so that the preloaded objects are directly attached to each bookmark object. You can have a look at a spike 
 I did in https://gist.github.com/tgxworld/885dc747c53a50b67970a32a76087deb

Reply:
Ah ok I know what the problem is now. Please ignore my suggestion to change `select_type` to `where` because if we're switching to the preloader pattern, we want to only load the query once and filter against the results array. Sorry for the confusion.